django/db/models/fields/composite.py

```diff
@@ -0,0 +1,146 @@
+from django.core.exceptions import FieldDoesNotExist
+from django.db.models import Field
+from django.db.models.fields.related_lookups import MultiColSource
+from django.db.models.lookups import Exact, In
+from django.db.models.signals import class_prepared
+from django.utils.functional import cached_property
+
+
+class TupleExact(Exact):
+    def get_prep_lookup(self):
+        if not isinstance(self.lhs, MultiColSource):
+            raise ValueError(
+                "The left-hand side of the 'exact' lookup must be an instance of "
+                "MultiColSource"
+            )
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(f"'{self.lhs.field.name}' must be a tuple or a list")
+        if len(self.lhs) != len(self.rhs):
+            raise ValueError(
+                f"'{self.lhs.field.name}' must have {len(self.lhs)} elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, WhereNode
+
+        cols = self.lhs.get_source_expressions()
+        exprs = [Exact(col, val) for col, val in zip(cols, self.rhs)]
+
+        return compiler.compile(WhereNode(exprs, connector=AND))
+
+
+class TupleIn(In):
+    def get_prep_lookup(self):
+        if not isinstance(self.lhs, MultiColSource):
+            raise ValueError(
+                "The left-hand side of the 'in' lookup must be an instance of "
+                "MultiColSource"
+            )
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(f"'{self.lhs.field.name}' must be a tuple or a list")
+        if not all(isinstance(vals, (tuple, list)) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lhs.field.name}' must be a collection of tuples or lists"
+            )
+        if not all(len(self.lhs) == len(vals) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lhs.field.name}' must have {len(self.lhs)} elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        exprs = []
+        cols = self.lhs.get_source_expressions()
+
+        for vals in self.rhs:
+            exprs.append(
+                WhereNode(
+                    [Exact(col, val) for col, val in zip(cols, vals)], connector=AND
+                )
+            )
```

Comment:
It'd be nice to eventually use [row level comparison instead](https://www.postgresql.org/docs/current/functions-comparisons.html#ROW-WISE-COMPARISON) for backends that support it.

Reply:
I could work on this in the next PR, however, I'm not sure if there's any real benefit to it other than generating nicer SQL.

Reply:
It got merged recently (https://github.com/django/django/pull/18404).